### ISSUE TITLE
MaxItemCount and FeedIterator.HasMoreResult retrieved less items than expected.

### ISSUE THREAD

**Issue by blankor1** [Azure/azure-cosmos-dotnet-v3#3963]
**Describe the bug**
We use QueryRequestOptions.MaxItemCount and FeedIterator.HasMoreResult to retrieve the specific number of items back like below sample code:
```
while ((requestOption.MaxItemCount <= 0 || list.Count < ro.MaxItemCount) && feedIterator.HasMoreResults)
                {
                    FeedResponse<T> feed = await feedIterator.ReadNextAsync();
                    foreach (var item in feed)
                    {
                        list.Add(item);
                    }
                }
```

This should retrieve the number of MaxItemCount back. It works for most cases and I test locally it also works welll. 

But in PROD env for some cases, we only retrieve small amount of item results back when there are far more item result that meet the query filter condition that exist in the database. **For our case, MaxItemCount is 150, item that meet the query filter condition is more than 50000, only retrieve 95 item back for lot of request.**

**To Reproduce**
Can't reprouce locally.

Set MaxItemCount = 150, data that meet the query condition is more than this

**Expected behavior**
Retrieve 150 items back

**Actual behavior**
only less than 150 items are retrieved back.

**Environment summary**
SDK Version: 3.31.0
OS Version Windows

**Comment by bartelink** [comment]
The server gets to control how many items to return - it may opt to respond before the limit is reached.
Sometimes this is due to processing time running out. Other times it can be due to exhausting of all the items on a given physical partition or partitions

Ultimately all the SDK can do is forward your limits, and pass on the result, so there's not much to be done or said in the context of this repo.

(I'd investigate the distribution of the items in question wrt how many physical partitions they are spread over etc; it can also be useful to look at the query analytics when you run the equivalent one interactively)

**Comment by blankor1** [comment]
@bartelink Thank you for your response! 

Our query is a cross partition query so even it exhausts all items on a physical partition, the FeedIterator.HasMoreResult should still be true if there are more items on other physical partition. But we just keep getting only part of results back. Is it possible that the FeedIterator.HasMoreResult = false while the query still have more data in the db?

Also, is there any way to know the distribution of item among physical partition? What can we do next step to avoid such case if we know the data distribution?

**Comment by bartelink** [comment]
I unfortunately don't know the answers to those questions (I avoid cross partition queries like the plague). Main things to check is that you're correctly constructing a cross-partition query (setting the right options, setting/not setting partition keys correctly etc).

But, I will say this: `SELECT` is not broken: There is no reason to suspect that there is a generalised problem with doing cross partition queries etc - this repo would have an open issue with a repro if there was a remotely interesting bug of that nature given the widespread usage of this SDK.

The distribution of items over physical partitions is something that's not exposed and/or something that you would/should want to know (it dynamically changes as physical partitions split and/or join).

The bottom line here is that question-issues about on this repo is not the best way to get this resolved - better to put a question on SO with details of exactly what you are trying, and work from there (feel free to paste a link here, but you can be sure that the same people that watch this repo and respond, also have searches set up on stack overflow too).

**Comment by ealsur** [comment]
@blankor1 As @bartelink correctly pointed out, this is a behavior from the service, the SDK has nothing to do with it.

MaxItemCount is passed as a header to the service and the service determines how many items to return, there are a lot of factors (see https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/query/pagination#understanding-query-executions) that can cause the query to return less items, the property says "Max" not equal or exact. It's a hint, not an enforcement.

> Is it possible that the FeedIterator.HasMoreResult = false while the query still have more data in the db?

No, if HasMoreResults is false, there is nothing more to read.

**Comment by adityasa** [comment]
Based on the (limited) code snippet above, the behavior is by design. If all results need to be fetched for a query, continuation token must be fully drained in the application code. Here's a resource that explains this in cosmos db:
https://learn.microsoft.com/en-us/azure/cosmos-db/nosql/query/pagination

Here's a code sample in .NET sdk that shows how to do this:
https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/Microsoft.Azure.Cosmos.Samples/Usage/Queries/Program.cs#L294

Specifically, please note that pagination loop for the purposes of draining the continuation token needs to be implemented based on HasMoreResults property and rather should not depend on MaxItemCount, # of documents returned in the response etc.

Hope this helps.

**Comment by lubyrex** [comment]
@adityasa Thanks for the response. However, looking at the examples I fail to understand the difference between @blankor1's code and the provided code sample in [the link](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/master/Microsoft.Azure.Cosmos.Samples/Usage/Queries/Program.cs#L294) you provided.

Let me try to rephrase the question to make it clearer:
I wrote a code snippet from the link you provided above and made some slight modifications:
```C#
            string queryText = "select * from c"; // cross partition query
            IList<Family> families1 = new List<Familiy>();
            using (FeedIterator<Family> query = container.GetItemQueryIterator<Family>(
                queryText,
                requestOptions: null)) // Setting request options null here
            {
                while (query.HasMoreResults)
                {
                    foreach (Family family in await query.ReadNextAsync())
                    {
                        families1.Add(family);
                    }
                }
            }
            
            IList<Family> families2 = new List<Familiy>();
            using (FeedIterator<Family> query = container.GetItemQueryIterator<Family>(
                queryText,
                requestOptions: new QueryRequestOptions  () {  MaxItemCount = 5 })) // Setting request options's MaxItemCount to some number
            {
                while (query.HasMoreResults)
                {
                    foreach (Family family in await query.ReadNextAsync())
                    {
                        families2.Add(family);
                    }
                }
            }
```

Questions:
1. Would families1 contain **all** the Familiy entities in the container given that the container has **multiple partitions**? Or do we need to manually manipulate `response.ContinuationToken` and call `container.GetItemQueryIterator` multiple times in order to iterate through **all** the Famility entities?
2. Would families1 and families2 lists contain the same results?

**Comment by ealsur** [comment]
Jumping in but probably @adityasa can call me out if I misspeak:

1. Yes. No need to do anything to account for multiple physical partitions.
2. Yes. The number of pages would just be different (assuming there are more than 5 families).

The idea of the ContinuationToken is to provide you a yield point for the query execution you can use, based on some business use case (for example, showing 1 page at a time on a UI, or consuming some data from one component, sending it to other for processing, then resuming once the other component signals it's done).

**Comment by blankor1** [comment]
@adityasa @ealsur We are basically doing things like:
```
using (FeedIterator<Family> query = container.GetItemQueryIterator<Family>(
                queryText,
                requestOptions: new QueryRequestOptions  () {  MaxItemCount = 150 }))
{
while (list.Count < 150 && feedIterator.HasMoreResults)
                {
                    FeedResponse<T> feed = await feedIterator.ReadNextAsync();
                    foreach (var item in feed)
                    {
                        list.Add(item);
                    }
                }
}
```
According to your description, this code should get at least 150 records back if there exist more than 150 records in the database since _feedIterator.HasMoreResults_ will be _true_. But what we actually facing is that: database has more than 50000 entities inside that meet the query filter, but only 95 entities are retrieved back continually for a while.

**Comment by adityasa** [comment]
Apologies for a late response.
@blankor1 there is no guarantee that above code will get a fixed (150 or some other number specified) number of records in each roundtrip.

MaxItemCount only indicates an upper limit on the number of documents to be returned in the query response to the client.
It doesn't (and cannot enforce) the exact count be met in the response. The number of documents returned by the service is determined by service alone and will be <= MaxItemCount (if specified by the client).

There are a number of other factors such as RUs and various service limits that affect the actual number of documents returned by the service. Here's a pointer to various service limits:
https://learn.microsoft.com/en-us/azure/cosmos-db/concepts-limits#sql-query-limits

Hope this helps.

**Comment by blankor1** [comment]
Thank you for your response. But the above code should guarantee that **at least** 150 records (>= 150) will be retrieved back if there exists such number of records that meet the query filter since we have a while loop condition on both list.Count <150 and feedIterator.HasMoreResults, right? 

Our query text is just like "select * from c where c.property = A order by c.ts ASC" so it seems doesn't hit the limit in the link you mentioned.

**Comment by adityasa** [comment]
@blankor1 apologies again for a late reponse. If you are still hitting this issue, kindly reopen this issue with latest repro. Query, sample console application for repro, activity ids etc would be most helpful.
